### PR TITLE
Splunk Plugin

### PR DESCRIPTION
**Pull request by amolkokane-cirus**
Added missing javadocs
Changed copyright to 2020

### CODE REVIEW

File: docs/Splunk-batchsource.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+**Username:** Login name for authentication to the Splunk API.
+
+**Password:** Password for authentication to the Splunk API.
+
+**Execution Mode:** Defines the behaviour for the Splunk Search.
+Valid values: (Blocking | Oneshot | Normal). Default is Normal.
+
+If set to Normal, runs an asynchronous search.
+
+If set to Blocking, returns the sid when the job is complete.
```
Comment (rmstar) [inline docs/Splunk-batchsource.md]: what is sid?

File: src/main/java/io/cdap/plugin/splunk/source/streaming/SplunkReceiver.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+
+          // reload current iterator
+          long iteratorPosition = searchIterator.getIteratorPosition();
+          searchIterator = new SplunkSearchIterator(config, searchId, iteratorPosition, allRecordsCount);
+          searchIterator.setIteratorPosition(iteratorPosition);
+        }
+      }
+    } catch (Exception e) {
+      String errorMessage = "Exception while receiving messages from Splunk";
+      throw new RuntimeException(errorMessage, e);
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/splunk/source/streaming/SplunkReceiver.java:94]: we should log retryable errors and try again. If we raise an exception, the receiver will be terminated and we won't get any more messages from the source.

File: src/main/java/io/cdap/plugin/splunk/source/streaming/SplunkReceiver.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+  @Override
+  public void onStop() {
+    // There is nothing we can do here as the thread calling receive()
+    // is designed to stop by itself if isStopped() returns false
+  }
+
+  private void receive() {
+    try {
+      long allRecordsCount = 0L;
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/splunk/source/streaming/SplunkReceiver.java:72]: allRecordsCount is always 0?

File: src/main/java/io/cdap/plugin/splunk/common/client/SplunkSearchClient.java [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    Job job = getJob(config);
+    while (!job.isDone()) {
+      if (!job.isReady()) {
+        sleep(timeToSleepMillis);
+        continue;
+      }
+      Object searchMode = config.getSearchArguments().get(SplunkStreamingSourceConfig.SEARCH_MODE);
+      if (SplunkStreamingSourceConfig.SEARCH_REALTIME.equals(searchMode)
+        || countOfRecords <= job.getResultCount()) {
+        return job.getSid();
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/splunk/common/client/SplunkSearchClient.java:152]: Please add a comment explaining why we return here if seach mode is `SEARCH_REALTIME` or `countOfRecords <= job.getResultCount()`

File: src/main/java/io/cdap/plugin/splunk/common/client/SplunkSearchClient.java [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+  }
+
+  /**
+   * Returns sample data for Splunk Search.
+   *
+   * @return List of events
+   * @throws IOException thrown if there are any issue with the I/O operations.
+   */
+  public List<Map<String, String>> getSample() throws IOException {
+    long countOfSamples = 100L;
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/splunk/common/client/SplunkSearchClient.java:86]: why 100 samples vs some other number, say 1?

File: src/main/java/io/cdap/plugin/splunk/sink/batch/SplunkBatchSink.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+  public SplunkBatchSink(SplunkBatchSinkConfig config) {
+    this.config = config;
+  }
+
+  @Override
+  public void configurePipeline(PipelineConfigurer pipelineConfigurer) {
+    FailureCollector collector = pipelineConfigurer.getStageConfigurer().getFailureCollector();
+    Schema inputSchema = pipelineConfigurer.getStageConfigurer().getInputSchema();
+    config.validate(collector, inputSchema);
```
Comment (rmstar) [inline src/main/java/io/cdap/plugin/splunk/sink/batch/SplunkBatchSink.java:61]: `inputSchema` will be null if previous stage's output schema is a macro.

**Review by rmstar — CHANGES_REQUESTED** [review]

File: docs/Splunk-batchsource.md (hunk 1 above)
Comment (amolkokane-cirus) [reply docs/Splunk-batchsource.md]: Done

File: src/main/java/io/cdap/plugin/splunk/source/streaming/SplunkReceiver.java (hunk 2 above)
Comment (amolkokane-cirus) [reply src/main/java/io/cdap/plugin/splunk/source/streaming/SplunkReceiver.java:94]: TODO

File: src/main/java/io/cdap/plugin/splunk/source/streaming/SplunkReceiver.java (hunk 3 above)
Comment (amolkokane-cirus) [reply src/main/java/io/cdap/plugin/splunk/source/streaming/SplunkReceiver.java:72]: TODO

File: src/main/java/io/cdap/plugin/splunk/common/client/SplunkSearchClient.java (hunk 4 above)
Comment (amolkokane-cirus) [reply src/main/java/io/cdap/plugin/splunk/common/client/SplunkSearchClient.java:152]: TODO

File: src/main/java/io/cdap/plugin/splunk/common/client/SplunkSearchClient.java (hunk 5 above)
Comment (amolkokane-cirus) [reply src/main/java/io/cdap/plugin/splunk/common/client/SplunkSearchClient.java:86]: TODO

File: src/main/java/io/cdap/plugin/splunk/sink/batch/SplunkBatchSink.java (hunk 6 above)
Comment (amolkokane-cirus) [reply src/main/java/io/cdap/plugin/splunk/sink/batch/SplunkBatchSink.java:61]: TODO

File: src/main/java/io/cdap/plugin/splunk/common/client/SplunkSearchClient.java (hunk 4 above)
Comment (rmstar) [reply src/main/java/io/cdap/plugin/splunk/common/client/SplunkSearchClient.java:152]: Please add a TODO comment in the code.

Something like:
// TODO: CDAP-xxx: explain why we return here if seach mode is SEARCH_REALTIME or countOfRecords <= job.getResultCount()

Similar comment for other TODOs

**Review by rmstar — APPROVED** [review]
lgtm after fixing TODO comments.

File: src/main/java/io/cdap/plugin/splunk/source/streaming/SplunkReceiver.java (hunk 2 above)
Comment (shiva-cirus) [reply src/main/java/io/cdap/plugin/splunk/source/streaming/SplunkReceiver.java:94]: Code updated and created JIRA - https://issues.cask.co/browse/PLUGIN-360

File: src/main/java/io/cdap/plugin/splunk/source/streaming/SplunkReceiver.java (hunk 3 above)
Comment (shiva-cirus) [reply src/main/java/io/cdap/plugin/splunk/source/streaming/SplunkReceiver.java:72]: Code updated and JIRA ticket created https://issues.cask.co/browse/PLUGIN-361

File: src/main/java/io/cdap/plugin/splunk/common/client/SplunkSearchClient.java (hunk 4 above)
Comment (shiva-cirus) [reply src/main/java/io/cdap/plugin/splunk/common/client/SplunkSearchClient.java:152]: Code updated and JIRA ticket created - https://issues.cask.co/browse/PLUGIN-362

File: src/main/java/io/cdap/plugin/splunk/common/client/SplunkSearchClient.java (hunk 5 above)
Comment (shiva-cirus) [reply src/main/java/io/cdap/plugin/splunk/common/client/SplunkSearchClient.java:86]: Code updated and JIRA ticket created.  https://issues.cask.co/browse/PLUGIN-363.

File: src/main/java/io/cdap/plugin/splunk/sink/batch/SplunkBatchSink.java (hunk 6 above)
Comment (shiva-cirus) [reply src/main/java/io/cdap/plugin/splunk/sink/batch/SplunkBatchSink.java:61]: Code updated with TODO and https://issues.cask.co/browse/PLUGIN-364

**Comment by shiva-cirus** [comment]
@rmstar  Code updated with all TODO

File: src/main/java/io/cdap/plugin/splunk/source/streaming/SplunkReceiver.java (hunk 2 above)
Comment (rmstar) [reply src/main/java/io/cdap/plugin/splunk/source/streaming/SplunkReceiver.java:94]: Add JIRA id in the comment.

**Review by rmstar — APPROVED** [review]

**Pull request merged**